### PR TITLE
refactor ImportTypesOfCancers

### PR DESCRIPTION
**Pull request by sheridancbio**
Describe changes proposed in this pull request:
- before wiping the type_of_cancer table, the parsing of the input file now must first complete successfully
- encapsulate two main functions:
    - parseCancerTypesFromFile()
    - writeRecordsToDatabase()
- display detailed information if input file parse errors occur (another encapsulated function)

# Checks
- [ ] Runs on heroku
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java [hunk 1]
```diff
@@ -32,74 +32,105 @@
...
-            String typeOfCancerId = tokens[0].trim();
-            //if not clobber, then existing cancer types should be skipped:
-            if (!clobber && DaoTypeOfCancer.getTypeOfCancerById(typeOfCancerId.toLowerCase()) != null ) {
-            	ProgressMonitor.logWarning("Cancer type with id '" + typeOfCancerId + "' already exists. Skipping.");
+        while (scanner.hasNextLine()) {
+            String nextLine = scanner.nextLine();
+            String[] fields = nextLine.split("\t", -1);
+            if (fields.length != 4) {
+                throwBadColumnCountException(file, nextLine, fields);
             }
```
Comment (inodb) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java:96]: @sheridancbio maybe good to add expectedFieldCount as argument to this function and move the `fields.length!=4` to be inside the function. Could rename the function to e.g. `checkColumnCount`. Currently we still have `4` and `four` in two places which is what caused the inconsistent message before

**Review by inodb — APPROVED** [review]
Thanks! Looks good to me - maybe nice to fix the issue with the four columns requirement being hardcoded in multiple places as well

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java (hunk 1 above)
Comment (sheridancbio) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java:96]: ok, will do

**Review by inodb — APPROVED** [review]
Great - thanks!

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java [hunk 2]
```diff
@@ -32,74 +32,107 @@
...
+            typeOfCancer.setShortName(typeOfCancerId);
+            typeOfCancer.setParentTypeOfCancerId(fields[3].trim().toLowerCase());
+            typeOfCancerList.add(typeOfCancer);
+        }
+        return typeOfCancerList;
+    }
+
+    private static void throwExceptionIfColumnCountIsWrong(File file, String nextLine, String[] fields, int expectedDatafileColumnCount) throws IOException {
+        if (fields.length != expectedDatafileColumnCount) {
+            String msg = String.format("Cancer type file '%s' contains a line which is not %d-column tab-delimited." +
```
Comment (jagnathan) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java:111]: the String.format to create String msg has 4 parameters, but only 3 have been passed. It requires expectedDatafileColumnCount also as a parameter.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java [hunk 3]
```diff
@@ -32,74 +32,107 @@
...
+        }
+    }
+
+    private static void writeRecordsToDatabase(List<TypeOfCancer> typeOfCancerList, boolean clobber) throws IOException, DaoException {
+        int numNewCancerTypes = 0;
+        for (TypeOfCancer typeOfCancer : typeOfCancerList) {
+            if (!clobber && typeOfCancerExistsInDatabase(typeOfCancer)) {
+                ProgressMonitor.logWarning(String.format("Cancer type with id '%s' already exists. Skipping.", typeOfCancer.getTypeOfCancerId()));
+            } else {
+                DaoTypeOfCancer.addTypeOfCancer(typeOfCancer);
```
Comment (ao508) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java:127]: will an error be thrown if clobber = true and the type of cancer already exists?

Does an update occur to the existing entry?

Comment (ao508) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java:127]: note: if clobber == true then the db table should already be wiped so this scenario shouldnt be possible?

**Review by ao508 — APPROVED** [review]

**Review by jagnathan — APPROVED** [review]

**Comment by ao508** [comment]
LGTM although the end to end tests failed due to some element not existing anymore `".Select-option=Pvalues of GSVA scores on oncogenic signatures gene sets" `

**Pull request merged**